_drafts/making-backend-scalable.markdown

```diff
@@ -0,0 +1,130 @@
+---
+layout: post
+title: "Making Autolab's Backend Scalable"
+author: Ilter Canberk
+
+tags: [tango, redis, scalability]
+
+---
+
+[Tango](https://github.com/autolab/Tango) is a stand-alone, RESTful service that Autolab uses as the back-end for autograding. Tango receives grading jobs from Autolab's front-end, adds them in a job queue, assigns them to available containers for grading and shepherds the job throuout the process. In its early days, Tango was mostly used for jobs that ran in under 5 seconds. However, over the recent semesters, Autolab has grown to host classes like Distributed Systems, Machine Learning. Some of these classes have assessments that are computationally intensive on large datasets and contains bigger files.
+
+
+![Tango's initial Architecture]({{site-url}}/assets/redis1.png)
+_Diagram 1: Initial Tango architecture with in-memory job queue_
+
+
+The existing monolithic model with a single process and an in-memory queue was posing multiple problems:
+
+* Running a web server that handled large file uploads concurrently along with a job manager that dispatched jobs from the job queue to new threads made the process prone to crashes.
+* Crashes caused tall the jobs stored in the in-memory queue to disappear. 
+* After a crash, all the containers running a job would suddenly be left in a so-called limbo state because they are no longer managed by any Tango process.
+
+
+Therefore we decided to improve our architecture by switching to multi-process model with a persistent queue.
+
+
+## Persistent Memory Model using Redis ##
+
+The in-memory job queue was the barrier to making the system robust. We want to store jobs on an independent system and make it *persistent*
+
+![Tango with Persistent Memory Model]({{site-url}}/assets/redis2.png)
+_Diagram 2: Tango with Persistent Memory Model_
+
+In this persistent memory model, even if Tango restarts it will keep the same job queue. We could even spin up multiple instances of Tango and they can all work concurrently, sharing the same queue (with a couple problems which I will mention in the next section).
+
+### Choice of Redis ###
+
+We started looking into possible solutions: traditional databases like MySQL, relational databases such as MongoDB and messaging queues such as RabbitMQ, Redis. Looking at the amount of data and the level of nestedness, we decided that we don't need a database with full-range of functionality and robustness. We needed something fast and simple to use. The implementation of our job queue is in the form of a Python dictionary, therefore, a fast key-value store like _Redis_ seemed like the right fit. Setting it up and playing around with it was a breeze, therefore we decided to go with this option.
+
+We are planning to add modules to support the other options in the future. More details about this will be give in the next sub-section.
```

Comment:
You can probably just remove this sentence, because it's unclear what "modules" you want to add, and you're going to talk about it in the next section anyway.